### ISSUE TITLE
[BUG] Does falco support private dictionaries?

### ISSUE THREAD

**Issue by ponkio-o** [ysugimoto/falco#141]
**Describe the problem**
We are using Private dictionaries to store credentials.
https://docs.fastly.com/en/guides/working-with-dictionaries-using-the-web-interface#private-dictionaries

**VCL code that cause the problem**
First, write code to read credentials stored in Private dictionaries. `aws_credentials` is private dictionary.
(See. https://docs.fastly.com/en/guides/amazon-s3#using-an-amazon-s3-private-bucket)

```vcl
sub vcl_miss {
#FASTLY miss

  call s3_request;

  return(fetch);
}

...

// @scope: pass, miss
sub s3_request {
  declare local var.awsAccessKey STRING;
  declare local var.awsSecretKey STRING;
  declare local var.awsS3Bucket STRING;
  declare local var.awsRegion STRING;
  declare local var.canonicalHeaders STRING;
  declare local var.signedHeaders STRING;
  declare local var.canonicalRequest STRING;
  declare local var.canonicalQuery STRING;
  declare local var.stringToSign STRING;
  declare local var.dateStamp STRING;
  declare local var.signature STRING;
  declare local var.scope STRING;

  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID");
  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY");
  set var.awsRegion = "ap-northeast-1";
...
```

Next, run `falco` and "Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400" is displayed.
```console
$ falco -r -I . ./custom_main.vcl
Remote option supplied. Fetch snippets from Fastly.
Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400
...omitted

🔥 [ERROR] undefined variable "aws_credentials"
in /path/to/custom_main.vcl at line 131, position 39
 130|
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID");
                                           ^^^^^^^^^^^^^^^
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY");

🔥 [ERROR] function table.lookup wants argument 1st as TABLE but applies NULL (function/argument-type)
in /path/to/custom_main.vcl at line 131, position 38
 130|
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID");
                                          ^
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY");
See document in detail: https://developer.fastly.com/reference/vcl/functions/table/table-lookup/

🔥 [ERROR] undefined variable "aws_credentials"
in /path/to/custom_main.vcl at line 132, position 39
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID");
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY");
                                           ^^^^^^^^^^^^^^^
 133|  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets");

...omitted
```

**Expected behavior**
I expected these errors are ignoring. Because, Private dictionaries can't read items.

> ・You cannot view items that belong to a private dictionary via the API.

see. [Limitations and considerations](https://docs.fastly.com/en/guides/working-with-dictionaries-using-the-web-interface#limitations-and-considerations)

**Desktop (please complete the following information):**
 - OS: Darwin/amd64
 - Shell: zsh

**Additional context**
I tried to use `// falco-ignore` but same behavior.
https://github.com/ysugimoto/falco/pull/140

**Comment by ysugimoto** [comment]
@ponkio-o

Thanks for reporting this issue.
I  haven't been careful about the private edge dictionary, I'll look into it with reproducing.

And just one thing,

```
  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID"); // falco-ignore
```

is it still raising the error?

**Comment by ponkio-o** [comment]
@ysugimoto 

Thank you for your mention quickly!

> And just one thing,
> set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID"); // falco-ignore
> is it still raising the error?

Yes.  I tried `// falco-ignore` and `// falco-ignore-start` / `// falco-ignore-end` but occured an error.
Actual result are below.

```console
$ falco -r -I . ./custom_main.vcl
Remote option supplied. Fetch snippets from Fastly.
Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400
🔥 [ERROR] undefined variable "aws_credentials"
in /path/to/custom_main.vcl at line 131, position 39
 130|
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID"); // falco-ignore
                                           ^^^^^^^^^^^^^^^
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY"); // falco-ignore

🔥 [ERROR] function table.lookup wants argument 1st as TABLE but applies NULL (function/argument-type)
in /path/to/custom_main.vcl at line 131, position 38
 130|
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID"); // falco-ignore
                                          ^
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY"); // falco-ignore
See document in detail: https://developer.fastly.com/reference/vcl/functions/table/table-lookup/

🔥 [ERROR] undefined variable "aws_credentials"
in /path/to/custom_main.vcl at line 132, position 39
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID"); // falco-ignore
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY"); // falco-ignore
                                           ^^^^^^^^^^^^^^^
 133|  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets"); // falco-ignore

🔥 [ERROR] function table.lookup wants argument 1st as TABLE but applies NULL (function/argument-type)
in /path/to/custom_main.vcl at line 132, position 38
 131|  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID"); // falco-ignore
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY"); // falco-ignore
                                          ^
 133|  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets"); // falco-ignore
See document in detail: https://developer.fastly.com/reference/vcl/functions/table/table-lookup/

🔥 [ERROR] undefined variable "s3_bucket_names"
in /path/to/custom_main.vcl at line 133, position 38
 132|  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY"); // falco-ignore
 133|  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets"); // falco-ignore
                                          ^^^^^^^^^^^^^^^
 134|  set var.awsRegion = "ap-northeast-1";

🔥 5 errors, ❗️ 0 warnings, 🔈 2 infos.
```

What concerns me is the error in `Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400`.

We have two dictionaries in this Fastly Service.

- s3_bucket_names
	- normal
- aws_credentials
	- private

![image](https://user-images.githubusercontent.com/29038315/234563094-6601a045-8a14-4701-9270-766725c2c658.png)

**Comment by ysugimoto** [comment]
@ponkio-o Thank you for explaining deeply!
And sorry I've mistaken for v0.20.0 release, so could you [download latest one](https://github.com/ysugimoto/falco/releases/tag/v0.20.0) and try it? It should works for ignoing feature.

Anyway, private edge dictionary behavior seems a bug so I'll try to fix it 😅

**Comment by ponkio-o** [comment]
I ran command in the new binary, so it's correctly!! 😄 Thank you!
```console
$ falco -r -I . ./custom_main.vcl
Remote option supplied. Fetch snippets from Fastly.
Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400
🔥 0 errors, ❗️ 0 warnings, 🔈 2 infos.
VCL looks fine ✨  And we suggested some informations to vcl get more accuracy 👍
To see informations detail, run command with -vv option.
```

I wrote earlier that it's displayed error when contains private dictionary in VCL. 
https://github.com/ysugimoto/falco/issues/141#issuecomment-1523272212

Specifically, the below case.
```vcl
...
  // falco-ignore-start
  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID");
  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY");
  // falco-ignore-end

  # s3_bucket_names is not private dictionary
  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets");
...
```

```console
$ falco -r -I . ./custom_main.vcl
Remote option supplied. Fetch snippets from Fastly.
Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400
🔥 [ERROR] undefined variable "s3_bucket_names"
in /path/to/custom_main.vcl at line 135, position 38
 134|  // falco-ignore-end
 135|  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets");
                                          ^^^^^^^^^^^^^^^
 136|  set var.awsRegion = "ap-northeast-1";

🔥 [ERROR] function table.lookup wants argument 1st as TABLE but applies NULL (function/argument-type)
in /path/to/custom_main.vcl at line 135, position 37
 134|  // falco-ignore-end
 135|  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets");
                                         ^
 136|  set var.awsRegion = "ap-northeast-1";
See document in detail: https://developer.fastly.com/reference/vcl/functions/table/table-lookup/

🔥 2 errors, ❗️ 0 warnings, 🔈 2 infos.
```

However, this problem can be avoided by using ignore feature, which seems to work fine for a whille.

```vcl
...
  // falco-ignore-start
  set var.awsAccessKey = table.lookup(aws_credentials, "AWS_ACCESS_KEY_ID");
  set var.awsSecretKey = table.lookup(aws_credentials, "AWS_SECRET_ACCESS_KEY");
  // falco-ignore-end

  # s3_bucket_names is not private dictionary
  set var.awsS3Bucket = table.lookup(s3_bucket_names, "assets"); // falco-ignore
...
```

```console
$ falco -r -I . ./custom_main.vcl
Remote option supplied. Fetch snippets from Fastly.
Fetching Edge Dictionaries...Failed to get edge dictionaries API respond not 200 code: 400
🔥 0 errors, ❗️ 0 warnings, 🔈 2 infos.
VCL looks fine ✨  And we suggested some informations to vcl get more accuracy 👍
To see informations detail, run command with -vv option.
```

**Comment by ysugimoto** [comment]
Hi @ponkio-o 

I've fixed the problem you reported on #144.
I have never considered the private edge dictionary before, but I learned a lot from your report 😄. 

Thank you for bringing this issue and please bear with us while we prepare to release a new version.

**Comment by ponkio-o** [comment]
Thank you for fix it!!
I will wait release for new version 😄
